streamz/sources.py

```diff
@@ -731,3 +731,36 @@ def get_message_batch_cudf(kafka_params, topic, partition, keys, low, high, time
     finally:
         consumer.close()
     return gdf
+
+
+@Stream.register_api(staticmethod)
+class from_iterable(Source):
+    """ Emits items from an iterable.
+
+    Parameters
+    ----------
+    iterable: iterable
+        An iterable to emit messages from.
+
+    Examples
+    --------
+
+    >>> source = Stream.from_iterable(range(3))
+    >>> L = source.sink_to_list()
+    >>> source.start()
+    >>> L
+    [0, 1, 2]
+    """
+
+    def __init__(self, iterable, **kwargs):
+        super().__init__(ensure_io_loop=True, **kwargs)
```

Comment:
There is actually nothing async about this source - do we need to bother with a loop? If yes, then are there any sources or other nodes which *don't* need a loop?

Reply:
> There is actually nothing async about this source - do we need to bother with a loop?

I think we do: the rest of the loop is still asynchronous, and we need to run `_run` as a callback on an IOLoop, so it should be ensured. This source is not asynchronous, but the rest of the pipeline is, so we need to await on whatever's next. Sources are always at the start of a pipeline, so they _are_ the place to make sure that a loop exists.

I tried to think how this could be implemented without an IOLoop and without `gen.coroutine` and I couldn't figure out how.

Reply:
Btw, there are also async generators, and I don't think they are supported in tornado. We might want to support them here, but then `_run` won't be in line with the rest of the codebase, because it would be a built-in `async def` coroutine.

Reply:
I see that all the rest of the sources are all also async, which I hadn't realised. In a separate PR, we should refactor all the `start()` methods, which seem to be mostly identical.

>  because it would be a built-in async def coroutine

as a separate issue, I don't know that tornado is warranted in this code anymore, everything could be done with asyncio.

Reply:
> as a separate issue, I don't know that tornado is warranted in this code anymore, everything could be done with asyncio.

If you're comfortable mixing both in the codebase, I can rewrite this. But there's a lot still going on with tornado ioloop in stream internals, I don't think it can be just asyncio code in nodes without changing how the whole thing works, `self.loop` is still tornado.